services/api-server/src/simcore_service_api_server/api/routes/files.py

```diff
@@ -0,0 +1,109 @@
+import asyncio
+from textwrap import dedent
+from typing import List
+
+from fastapi import APIRouter, Depends, File, UploadFile, status
+from fastapi.exceptions import HTTPException
+from fastapi.responses import FileResponse, HTMLResponse
+
+from ..._meta import api_vtag
+from ...models.schemas.files import FileUploaded
+from ...modules.storage import StorageApi
+from ..dependencies.services import get_api_client
+from .files_faker import the_fake_impl
+
+router = APIRouter()
+
+
+## FILES ---------------
+
+
+@router.get("", response_model=List[FileUploaded])
+async def list_files():
+    # TODO: extend sear
+    return the_fake_impl.list_meta()
+
+
+async def list_files_impl(
+    _storage_client: StorageApi = Depends(get_api_client(StorageApi)),
+):
+    # TODO: pagination
+    # TODO: this is just a ping with retries
+    # TODO: extend :search see https://cloud.google.com/apis/design/custom_methods
+    # await storage_client.get("/")
+    raise NotImplementedError()
+
+
+@router.post(":upload", response_model=FileUploaded)
+async def upload_single_file(
+    file: UploadFile = File(...),
+):
+    metadata = await the_fake_impl.save(file)
+    return metadata
+
+
+async def upload_single_file_impl(
+    file: UploadFile = File(...),
+    _storage_client: StorageApi = Depends(get_api_client(StorageApi)),
+):
+    # TODO: FileResponse automatically computes etag. See how is done
+
+    # TODO: every file uploaded is sent to S3 and a link is returned
+    # TODO: every session has a folder. A session is defined by the access token
+    #
+
+    # TODO: this is just a ping with retries
+    ## await storage_client.get("/")
+    raise NotImplementedError()
+
+
+@router.post(":upload-multiple", response_model=List[FileUploaded])
+async def upload_multiple_files(files: List[UploadFile] = File(...)):
+    # TODO: every file uploaded is sent to S3 and a link is returned
```

Comment:
one question. is it not possible to have one /upload_files endpoint that takes both a single or a list of files using a Union?

Reply:
It is a good idea but  there are two problems: 1) it produces an error in generated OAS, 2) our client generator does not know how to handle that. 
As a compromise we could also have upload_files(files: List) only and uploading one file would be adding a single element list. What do you think?